tools/config/__init__.py

```diff
@@ -1092,17 +1092,38 @@ def _check_required_parameters(params):
                                       "' doesn't have a value")
 
     @staticmethod
-    def parameters_to_macros(params):
-        """ Encode the configuration parameters as C macro definitions.
+    def parameters_and_config_macros_to_macros(params, macros):
+        """ Return the macro definitions generated for a dictionary of
+        ConfigParameters and a dictionary of ConfigMacros (as returned by
+        get_config_data). The ConfigMacros override any matching macros set by
+        the ConfigParameters.
 
         Positional arguments:
         params - a dictionary mapping a name to a ConfigParameter
+        macros - a dictionary mapping a name to a ConfigMacro
 
-        Return: a list of strings that encode the configuration parameters as
-        C pre-processor macros
+        Return: a list of strings that are the C pre-processor macros
         """
-        return ['%s=%s' % (m.macro_name, m.value) for m in params.values()
-                if m.value is not None]
+        all_macros = {
+            p.macro_name: p.value for p in params.values() if p.value is not None
+        }
+
+        config_macros = {}
+        for macro in macros.values():
+            parts = macro.name.split("=", 1)
+            parts = [p.strip() for p in parts]
+            config_macros[parts[0]] = parts[1] if len(parts) > 1 else None
```

Comment:
Should this be `>= 1`

Reply:
Nope! I should really add comments here since this is a little opaque. This is dealing with directly specified macros in the configuration (entries like this: https://github.com/ARMmbed/mbed-cloud-client-example/blob/master/mbed_app.json#L3-L4). Those entries can be either of the form `MACRO=VALUE` or just `MACRO`. So this is basically saying if `=` is present, use the right hand side as the value, otherwise set it to `None` (indicating no value).

Reply:
@bridadan You could avoid having to explain yourself if you do the following:
```suggestion
            try:
                name, value = parts
                config_macros[name] = value
            except ValueError:
                name, = parts
                config_macros[name] = None
```